File: python/generator.py
import random

from maze import Maze


def binary_tree(w, h):
    m = Maze(w*2+1, h*2+1)
    for x in range(w):
        for y in range(h):
            mx = 2*x+1
            my = 2*y+1
            # Clear all cells.
            m.clear(mx, my)
            # Clear walls selectively.
            if (random.random() < 0.5  or x == w-1) and y != h - 1:
                # Clear below.
                m.clear(mx, my+1)
            elif x != w-1:
                # Clear right.
                m.clear(mx+1, my)
    return m

File: src/generator.ts
import { Observable, Subscriber } from "rxjs";
import { Cell, Maze } from "./maze";

// Returns an Observable of cleared Cells.
export function binaryTree(m: Maze): Observable<Cell> {
    const w = m.width;
    const h = m.height;
    return new Observable(subscriber => {
        const clear = (x: number, y: number) => {
            m.clear(x, y);
            subscriber.next(m.at(x, y));
        }
        for (let y = 1; y < h - 1; y += 2) {
            for (let x = 1; x < w - 1; x += 2) {
                clear(x, y);
                if (y < h - 2 && (Math.random() < 0.5 || x >= w - 2)) {
                    clear(x, y+1);
                } else if (x < w - 2) {
                    clear(x+1, y);
                }
            }
        }
        subscriber.complete();
    });
}
